### DISCUSSION TITLE
Manual Layout

### DISCUSSION THREAD

**Discussion by JasonBSteele** [likec4/likec4#343] [Ideas]
I would love to use LikeC4 as I believe the flexibility it provides for the modelling exceeds that of C4. However, with Structurizr I can use auto layout initially but then switch to manual to rearrange the positions to be more logical for what the diagram is trying to show. This might be moving Containers or Components that are primary or important services above those that are secondary or supporting.

**Comment by aelberg** [comment]
The https://template.likec4.dev/view/aws/editor site has an Editor Preview.  Is this a feature that is coming?

**Reply by JasonBSteele** [reply]
Well that looks promising!

**Comment by JasonBSteele** [comment]
It looks like the manual layout is coming along. In VS Code I can move the shapes around and a comment gets generated above the view code in the likec4 file. However, after building the site using `likec4 build` and then viewing it using `likec4 preview` the manual layout is no longer shown.

However, if I switch to Editor, my saved layout is shown. 

Of course, in Editor I can now also move around the shapes, but their position won't be saved due to it being a stateless website.

Is there any way to get the React view to show my manual layout?

Also, are there any options for controlling the web site that gets built? - I don't really need the drop-down of different renderers.

**Reply by JasonBSteele** [reply]
Hi @davydkov, any thoughts on this? Are we able to use different template sites from the default one that gets used for the builds? I need to be able to show the manual layout without forcing the user to select "Editor". Also, ideally I wouldn't have those rendering choices there at all.

**Comment by simonwfarrow** [comment]
Hi, I only noticed the 'comment' layout in vs code recently. This is great that you can save a manual layout, however I have many files and many views and the layout comments can get quite large; I was wondering if the layout comment would be better placed in a related 'hidden' file , perhaps a .layout file per dsl file in the workspace?

**Reply by JasonBSteele** [reply]
Yes there are definitely some issues with placing those comments above the views they represent: they can get big and as you edit the diagram VS Code jumps to the generated comment in the left pane losing your place and looking a little glitchy. This gets particularly bad when there is a bit of delay before the comment gets updated and you have started to edit the code. 

Also even though the comments can be collapsed in VS Code, they unfortunately expand again when you edit the diagram.

However, perhaps a decision was made to put it there to avoid some potential issues with using "shadow" files. By having the comment with the view you can presumably cut and paste the view and its layout to other places in the doc or even other files. If the layout is in a shadow file it can very easily get out of sync with the code. This frequently happen with Structurizr (even using explicit view IDs) and is one of the reasons I am here! :) 

I would prefer a shadow file - but not at the expense of sync issues.

**Reply by simonbrowndotje** [reply]
Just as an FYI - layout sync issues in Structurizr are generally the result of (1) not using explicit view keys or (2) renaming large quantities of elements in a single batch. The layout merge strategy is pluggable, so you could certainly write your own implementation to mirror what LikeC4 is doing here, which seems to be merging based upon the identifiers that are specified in the DSL. The trade-off of this approach is that (1) element layout is lost when a DSL identifier is changed (this happens for me on the playground, anyway) and (2) all elements must have an identifier. It's all trade-offs...

**Comment by simonwfarrow** [comment]
@JasonBSteele  good points, agree that an out of sync shadow file would not be ideal. I suppose another option could be to only update the view layout when the user requests i.e. add a button to the preview editor to save layout (or at least have it as an option to auto-save or not); this would resolve the flicking back and forth between windows while making lots of changes.
 
In addition, making the comment single line rather than multi line would make the file feel less cluttered I feel.

**Comment by wimdeblauwe-swmu** [comment]
How do you enable manual layouting locally? I am using the VS Code extension and I run `npx likec4 start` to preview. But I am unable to manually drag elements around like in the online playground.

**Reply by davydkov** [reply]
Hey @wimdeblauwe-swmu!

Do I understand you correctly? You want to modify the layout in the browser after running `npx likec4 start`?

It’s possible by adding “/editor” to the address, like http://localhost:5173/view/cloud/editor. However, note that any changes made are not saved (this feature is primarily used for development).

Previously, there was a button to navigate to the editor, but it also appeared in the static website, which was confusing.

The layout editor works as expected when you preview the diagram in VSCode

**Reply by wimdeblauwe-swmu** [reply]
Thanks! I did not find that information in the documentation.

**Comment by dilzio** [comment]
+1 for solid manual layout... this is really necessary to get to clear visual storytelling and really unlock the power of likec4 (which is amazing overall).    I've worked with the manual layout in VS Code and it's a pretty good start but super frustrating as it's really tough to remove control points from the connectors so when you move elements around, those points stay pinned and it's very difficult to fix them.

**Reply by dilzio** [reply]
sorry forgot to mention the main dealbreaker.  Altering any field on an element seems to trash the manual layout.  I added a description to one relationship and it trashed the entire layout I'd been working on for 45 minutes....Is there a way to prevent this?

**Reply by aboriskin** [reply]
Hi @dilzio. Any luck on this? I'm struggling with the same issue at the moment, and this is a necessary feature for me to continue with likec4

**Reply by ambition-consulting** [reply]
same here. this tool is awesome, but what when diagrams get a little more complicated, manual corrections become really important.

**Comment by yubrshen** [comment]
The manual layout looks promising. Here are some suggestions to improve: 
as shown in the screenshot from the preview with VS Code LikeC4 extension,
![image](https://github.com/user-attachments/assets/11e59392-52c5-4eac-af1c-c2a3a8e97f5a)

1. The positions of the links with the manually moved nodes might be adjusted for the closest connecting ports (from s -> n to e -> w, etc.) of the impacted nodes.
2. The changes of rank distance, and node distance in the editor ideally should be synched  with those reflected in the preview window.
3. Besides the supported layout Top-Bottom, could we have more layout options? 
For my case, as shown in the screenshots, for the nodes, A, B, C, D. I might want to have the options to specify the link orientation. For example, 
A -horizontal-> B
B -vertical-> C
C- horizontal-reverse -> D (placing D to the left of C)
(PlantUML might support some the above.)
If we support such flexibilities, we might save some of the challenges and needs for manual layout.

Using hash-code to capture the manual intervention is a clever design. But it is opaque of the semantics of the adjustment, against the idea of architecture as code (at least, for human readable code). Besides, those hash-code might not be adaptive to further changes.

Providing more heuristics through the orientations of the links might partially address the needs for manual layout. They are transparent of the intentions. Alternatively, we might express the instruction of the link orientations as part of the links properties to simplify the syntax for links. For example,

A -> B {orientation=horizontal}
B -> C {orientation=vertical}
C -> D {orientation=horizontal-reverse}
(placing D to the left of C)

We may have user configuration for the extent of complying those heuristics.

Again, thanks for the great work! It's almost my dream come true.

**Reply by yubrshen** [reply]
Alternatively, to support more declarative layout control, the approach of graphviz might be a source of inspiration. graphviz uses rank and ports of connectors to influence the layout. 
So far, graphviz seems provide the most robust declarative layout.

**Comment by yubrshen** [comment]
I haven't figured out how to select multiple nodes for applying alignment in "Manual layouting tools". 
I tried Ctrl-click, and Shift-click in Windows VS Code, it seems the behavior is not certain.
Where can I find some description or demonstration to them?
(I did figure out using drag and drop with the nodes, also discovered drag and drop the links by dragging the "handle" of a link, but 
I find dragging a link's handle seeming difficult to achieve desired results.)

**Reply by pavelpykhtin** [reply]
Hi, @yubrshen 

There are two ways to select multiple nodes:
1. Area selection with Shift + Drag
2. Add to selection with Ctrl + Click

Not sure when the second option was implemented. I'm almost sure this was not possible in 1.19, but now in 1.22.1 it works.

**Reply by yubrshen** [reply]
@pavelpykhtin I have tried. In the latest VS Code extension as of Oct. 28, 2025, Shift + Drag does not work for nodes selection. Ctrl + Click does. But I have tried all those "manual control" buttons such as "align in columns", "align in rows", "align left", etc. It seems that none of the alignment controls behave as promised according to my interpretations of those labels.

**Comment by Yash-xoxo** [comment]
Manual layout is definitely needed for complex diagrams. The auto-layout works great for simple cases but struggles with:
- Dense interconnections
- Specific flow directions (e.g., data pipeline left-to-right)
- Grouping by non-hierarchical concerns (teams, environments)

**Current workarounds:**

1. **Use layout hints:**
```c4
view {
    layout: LR  // left-to-right
    autoLayout {
        rankSep: 100
        nodeSep: 50
    }
}
```

2. **Manual positioning (if supported):**
```c4
element api {
    position: {x: 100, y: 200}
}
```

3. **Nested grouping to influence layout:**
```c4
container frontend_layer {
    element web
    element mobile
}
container backend_layer {
    element api
}
```

**What would be ideal:**

- Drag-and-drop positioning in the UI (save to `.c4` file)
- Lock specific elements while auto-laying others
- Grid/alignment guides
- Layout templates (flowchart, layered, radial)

From the screenshot, looks like you're dealing with a vertical flow that auto-layout isn't handling well. Try `layout: TB` with increased `rankSep` to spread it out more vertically.

Is this for the VSCode extension or the web viewer?
